examples/src/main/java/io/grpc/examples/manualflowcontrol/ManualFlowControlServer.java

```diff
@@ -0,0 +1,150 @@
+        // stream's flow control using the response stream's observer, but this is the way it is.
+        final ServerCallStreamObserver<HelloReply> serverCallStreamObserver =
+            (ServerCallStreamObserver<HelloReply>) responseObserver;
+        serverCallStreamObserver.disableAutoInboundFlowControl();
+        // Signal the request sender to send one message.
+        serverCallStreamObserver.request(1);
+
+        // Set up a back-pressure-aware producer for the response stream. The onReadyHandler will be invoked
+        // when the consuming side has enough buffer space to receive more messages.
+        //
+        // Note: the onReadyHandler is invoked by gRPC's internal thread pool. You can't block in this in
+        // method or deadlocks can occur.
+        serverCallStreamObserver.setOnReadyHandler(new Runnable() {
+          @Override
+          public void run() {
+            // Start generating values from where we left off on a non-gRPC thread.
+            pool.execute(new Runnable() {
+              @Override
+              public void run() {
+                // requestStream.isReady() will go false when the consuming side runs out of buffer space and
+                // signals to slow down with back-pressure.
+                while (serverCallStreamObserver.isReady() && !serverCallStreamObserver.isCancelled()) {
+                  if (!work.isEmpty()) {
+                    // Send more messages if there are more messages to send.
+                    String name = work.remove();
+                    if (name != COMPLETED) {
+                      // Simulate doing some work to generate the next value
+                      try {
+                        Thread.sleep(200);
+                      } catch (InterruptedException e) {
+                        e.printStackTrace();
+                      }
+
+                      String message = "Hello " + name;
+                      System.out.println("<-- " + message);
+                      HelloReply reply = HelloReply.newBuilder().setMessage(message).build();
+
+                      // Send a response.
+                      responseObserver.onNext(reply);
+                    } else {
+                      System.out.println("Done.");
+                      responseObserver.onCompleted();
+                    }
+                  }
+                }
+              }
+            });
+          }
+        });
+
+        // Give gRPC a StreamObserver it can write incoming requests into.
+        return new StreamObserver<HelloRequest>() {
+          @Override
+          public void onNext(HelloRequest request) {
+            // Process the request and send a response or an error.
+            try {
+              // Accept and enqueue the request.
+              String name = request.getName();
+              System.out.println("--> " + name);
+              work.add(name);
```

Comment:
I think this is the source of confusion:
> Because onReadyHandler is looping, the callback thread is blocked until isReady() == false, but isReady() can never be set to false because doing so requires the callback thread, which is stuck in a loop.

`isReady()` changes immediately. As soon as `onNext()` returns when sending a message, `isReady()` is up-to-date. If it didn't it would be pretty weak, like you are describing.

> Additionally, since the callback thread is stuck in a loop, no inbound messages are ingested. Ingestion is also handled by the callback thread.

That is on purpose. Since the server is doing some "processing" it should push-back on the client. It should only request more messages when it is ready to process them. Having an infinite queue of requests is counter-productive.